[D,lib/expression/node/OperatorNode.js,71]     var jsArgs = node.args.map(function (arg) {
[A,lib/expression/node/OperatorNode.js,81]     var jsArgs = map(node.args, function (arg) {
[C,lib/expression/node/OperatorNode.js,82]       return compile(arg, defs, args);
[C,lib/expression/node/OperatorNode.js,83]     });
[C,lib/expression/node/OperatorNode.js,84] 
[D,lib/expression/node/OperatorNode.js,75]     // FIXME: escape/validate fn
[D,lib/expression/node/OperatorNode.js,76]     return 'math.' + node.fn + '(' + jsArgs.join(', ') + ')';
[A,lib/expression/node/OperatorNode.js,85]     return 'math[' + stringify(node.fn) + '](' + join(jsArgs, ', ') + ')';
[C,lib/expression/node/OperatorNode.js,86]   }
[C,lib/expression/node/OperatorNode.js,87] 
[C,lib/expression/node/OperatorNode.js,88]   // register the compile function
[C,lib/expression/node/OperatorNode.js,89]   register(OperatorNode.prototype.type, compileOperatorNode);
[C,lib/expression/node/OperatorNode.js,90] 
[C,lib/expression/node/OperatorNode.js,91]   /**
[C,lib/expression/node/OperatorNode.js,92]    * Execute a callback for each of the child nodes of this node
[C,lib/expression/node/OperatorNode.js,93]    * @param {function(child: Node, path: string, parent: Node)} callback
[C,lib/expression/node/OperatorNode.js,94]    */
[C,lib/expression/node/OperatorNode.js,95]   OperatorNode.prototype.forEach = function (callback) {
[C,lib/expression/node/OperatorNode.js,96]     for (var i = 0; i < this.args.length; i++) {
[C,lib/expression/node/OperatorNode.js,97]       callback(this.args[i], 'args[' + i + ']', this);
[C,lib/expression/node/OperatorNode.js,98]     }
[C,lib/expression/node/OperatorNode.js,99]   };
[C,lib/expression/node/OperatorNode.js,100] 
@@ -1,20 +1,22 @@
[C,lib/expression/node/compile.js,1] // the compile functions which compile a Node into JavaScript are not
[C,lib/expression/node/compile.js,2] // exposed as class methods for security reasons to prevent being able to
[C,lib/expression/node/compile.js,3] // override them or create fake Nodes. Instead, only compile functions of
[C,lib/expression/node/compile.js,4] // registered nodes can be executed
[C,lib/expression/node/compile.js,5] 
[A,lib/expression/node/compile.js,6] var hasOwnProperty = require('../../utils/object').hasOwnProperty;
[A,lib/expression/node/compile.js,7] 
[C,lib/expression/node/compile.js,8] function factory () {
[C,lib/expression/node/compile.js,9]   // map with node type as key and compile functions as value
[C,lib/expression/node/compile.js,10]   var compileFunctions = {}
[C,lib/expression/node/compile.js,11] 
[C,lib/expression/node/compile.js,12]   /**
[C,lib/expression/node/compile.js,13]    * Register a compile function for a node
[C,lib/expression/node/compile.js,14]    * @param {string} type
[C,lib/expression/node/compile.js,15]    * @param {function} compileFunction
[C,lib/expression/node/compile.js,16]    *                      The compile function, invoked as
[C,lib/expression/node/compile.js,17]    *                      compileFunction(node, defs, args)
[C,lib/expression/node/compile.js,18]    */
[C,lib/expression/node/compile.js,19]   function register(type, compileFunction) {
[C,lib/expression/node/compile.js,20]     if (compileFunctions[type] === undefined) {
[C,lib/expression/node/compile.js,21]       compileFunctions[type] = compileFunction;
[C,lib/expression/node/compile.js,22]     }
@@ -24,36 +26,36 @@ function factory () {
[C,lib/expression/node/compile.js,26]   }
[C,lib/expression/node/compile.js,27] 
[C,lib/expression/node/compile.js,28]   /**
[C,lib/expression/node/compile.js,29]    * Compile a Node into JavaScript
[C,lib/expression/node/compile.js,30]    * @param {Node} node
[C,lib/expression/node/compile.js,31]    * @param {Object} defs     Object which can be used to define functions
[C,lib/expression/node/compile.js,32]    *                          or constants globally available for the compiled
[C,lib/expression/node/compile.js,33]    *                          expression
[C,lib/expression/node/compile.js,34]    * @param {Object} args     Object with local function arguments, the key is
[C,lib/expression/node/compile.js,35]    *                          the name of the argument, and the value is `true`.
[C,lib/expression/node/compile.js,36]    *                          The object may not be mutated, but must be
[C,lib/expression/node/compile.js,37]    *                          extended instead.
[C,lib/expression/node/compile.js,38]    * @return {string} Returns JavaScript code
[C,lib/expression/node/compile.js,39]    */
[C,lib/expression/node/compile.js,40]   function compile (node, defs, args) {
[D,lib/expression/node/compile.js,39]     if (Object.hasOwnProperty.call(compileFunctions, node.type)) {
[A,lib/expression/node/compile.js,41]     if (hasOwnProperty(compileFunctions, node.type)) {
[C,lib/expression/node/compile.js,42]       var compileFunction = compileFunctions[node.type];
[C,lib/expression/node/compile.js,43]       return compileFunction(node, defs, args);
[C,lib/expression/node/compile.js,44]     }
[C,lib/expression/node/compile.js,45]     else if (typeof node._compile === 'function' &&
[D,lib/expression/node/compile.js,44]         !Object.hasOwnProperty.call(node, '_compile')) {
[A,lib/expression/node/compile.js,46]         !hasOwnProperty(node, '_compile')) {
[C,lib/expression/node/compile.js,47]       // Compatibility for CustomNodes
[C,lib/expression/node/compile.js,48]       // TODO: this is a security risk, change it such that you have to register CustomNodes separately in math.js, like math.expression.node.register(MyCustomNode)
[C,lib/expression/node/compile.js,49]       return node._compile(defs, args);
[C,lib/expression/node/compile.js,50]     }
[C,lib/expression/node/compile.js,51]     else {
[C,lib/expression/node/compile.js,52]       throw new Error('Cannot compile node: unknown type "' + node.type + '"');
[C,lib/expression/node/compile.js,53]     }
[C,lib/expression/node/compile.js,54]   }
[C,lib/expression/node/compile.js,55] 
[C,lib/expression/node/compile.js,56]   return {
[C,lib/expression/node/compile.js,57]     register: register,
[C,lib/expression/node/compile.js,58]     compile: compile
[C,lib/expression/node/compile.js,59]   }
[C,lib/expression/node/compile.js,60] }
[C,lib/expression/node/compile.js,61] 
@@ -0,0 +1,20 @@
[A,lib/expression/node/utils/getUniqueArgumentsName.js,1] /**
[A,lib/expression/node/utils/getUniqueArgumentsName.js,2]  * Get a unique name for a arguments to store in defs
[A,lib/expression/node/utils/getUniqueArgumentsName.js,3]  * @param {Object} defs
[A,lib/expression/node/utils/getUniqueArgumentsName.js,4]  * @return {string} A string like 'args1', 'args2', ...
[A,lib/expression/node/utils/getUniqueArgumentsName.js,5]  * @private
[A,lib/expression/node/utils/getUniqueArgumentsName.js,6]  */
[A,lib/expression/node/utils/getUniqueArgumentsName.js,7] function getUniqueArgumentsName (defs) {
[A,lib/expression/node/utils/getUniqueArgumentsName.js,8]   var argsName;
[A,lib/expression/node/utils/getUniqueArgumentsName.js,9]   var i = 0;
[A,lib/expression/node/utils/getUniqueArgumentsName.js,10] 
[A,lib/expression/node/utils/getUniqueArgumentsName.js,11]   do {
[A,lib/expression/node/utils/getUniqueArgumentsName.js,12]     argsName = 'args' + i;
[A,lib/expression/node/utils/getUniqueArgumentsName.js,13]     i++;
[A,lib/expression/node/utils/getUniqueArgumentsName.js,14]   }
[A,lib/expression/node/utils/getUniqueArgumentsName.js,15]   while (argsName in defs);
[A,lib/expression/node/utils/getUniqueArgumentsName.js,16] 
[A,lib/expression/node/utils/getUniqueArgumentsName.js,17]   return argsName;
[A,lib/expression/node/utils/getUniqueArgumentsName.js,18] }
[A,lib/expression/node/utils/getUniqueArgumentsName.js,19] 
[A,lib/expression/node/utils/getUniqueArgumentsName.js,20] module.exports = getUniqueArgumentsName;
@@ -1,27 +1,26 @@
[C,lib/expression/transform/map.transform.js,1] 'use strict';
[C,lib/expression/transform/map.transform.js,2] 
[C,lib/expression/transform/map.transform.js,3] var maxArgumentCount = require('../../utils/function').maxArgumentCount;
[C,lib/expression/transform/map.transform.js,4] 
[C,lib/expression/transform/map.transform.js,5] /**
[C,lib/expression/transform/map.transform.js,6]  * Attach a transform function to math.map
[C,lib/expression/transform/map.transform.js,7]  * Adds a property transform containing the transform function.
[C,lib/expression/transform/map.transform.js,8]  *
[C,lib/expression/transform/map.transform.js,9]  * This transform creates a one-based index instead of a zero-based index
[C,lib/expression/transform/map.transform.js,10]  */
[C,lib/expression/transform/map.transform.js,11] function factory (type, config, load, typed) {
[D,lib/expression/transform/map.transform.js,12]   var map = load(require('../../function/matrix/map'));
[C,lib/expression/transform/map.transform.js,12]   var matrix = load(require('../../type/matrix/function/matrix'));
[C,lib/expression/transform/map.transform.js,13] 
[C,lib/expression/transform/map.transform.js,14]   return typed('max', {
[C,lib/expression/transform/map.transform.js,15]     'Array, function': function (x, callback) {
[C,lib/expression/transform/map.transform.js,16]       return _map(x, callback, x);
[C,lib/expression/transform/map.transform.js,17]     },
[C,lib/expression/transform/map.transform.js,18] 
[C,lib/expression/transform/map.transform.js,19]     'Matrix, function': function (x, callback) {
[C,lib/expression/transform/map.transform.js,20]       return matrix(_map(x.valueOf(), callback, x));
[C,lib/expression/transform/map.transform.js,21]     }
[C,lib/expression/transform/map.transform.js,22]   });
[C,lib/expression/transform/map.transform.js,23] }
[C,lib/expression/transform/map.transform.js,24] 
[C,lib/expression/transform/map.transform.js,25] /**
[C,lib/expression/transform/map.transform.js,26]  * Map for a multi dimensional array. One-based indexes
@@ -406,21 +406,48 @@ exports.flatten = function(array) {
[C,lib/utils/array.js,406]   }
[C,lib/utils/array.js,407]   var flat = [];
[C,lib/utils/array.js,408] 
[C,lib/utils/array.js,409]   array.forEach(function callback(value) {
[C,lib/utils/array.js,410]     if (Array.isArray(value)) {
[C,lib/utils/array.js,411]       value.forEach(callback);  //traverse through sub-arrays recursively
[C,lib/utils/array.js,412]     }
[C,lib/utils/array.js,413]     else {
[C,lib/utils/array.js,414]       flat.push(value);
[C,lib/utils/array.js,415]     }
[C,lib/utils/array.js,416]   });
[C,lib/utils/array.js,417] 
[C,lib/utils/array.js,418]   return flat;
[C,lib/utils/array.js,419] };
[C,lib/utils/array.js,420] 
[A,lib/utils/array.js,421] /**
[A,lib/utils/array.js,422]  * A safe map
[A,lib/utils/array.js,423]  * @param {Array} array
[A,lib/utils/array.js,424]  * @param {function} callback
[A,lib/utils/array.js,425]  */
[A,lib/utils/array.js,426] exports.map = function (array, callback) {
[A,lib/utils/array.js,427]   return Array.prototype.map.call(array, callback);
[A,lib/utils/array.js,428] }
[A,lib/utils/array.js,429] 
[A,lib/utils/array.js,430] /**
[A,lib/utils/array.js,431]  * A safe forEach
[A,lib/utils/array.js,432]  * @param {Array} array
[A,lib/utils/array.js,433]  * @param {function} callback
[A,lib/utils/array.js,434]  */
[A,lib/utils/array.js,435] exports.forEach = function (array, callback) {
[A,lib/utils/array.js,436]   Array.prototype.forEach.call(array, callback);
[A,lib/utils/array.js,437] }
[A,lib/utils/array.js,438] 
[A,lib/utils/array.js,439] /**
[A,lib/utils/array.js,440]  * A safe join
[A,lib/utils/array.js,441]  * @param {Array} array
[A,lib/utils/array.js,442]  * @param {string} separator
[A,lib/utils/array.js,443]  */
[A,lib/utils/array.js,444] exports.join = function (array, separator) {
[A,lib/utils/array.js,445]   return Array.prototype.join.call(array, separator);
[A,lib/utils/array.js,446] }
[A,lib/utils/array.js,447] 
[C,lib/utils/array.js,448] /**
[C,lib/utils/array.js,449]  * Test whether an object is an array
[C,lib/utils/array.js,450]  * @param {*} value
[C,lib/utils/array.js,451]  * @return {boolean} isArray
[C,lib/utils/array.js,452]  */
[C,lib/utils/array.js,453] exports.isArray = Array.isArray;
@@ -50,30 +50,40 @@ function setSafeProperty (object, prop, value) {
[C,lib/utils/customs.js,50]       // override when the property is defined on the object itself.
[C,lib/utils/customs.js,51]       // don't allow overriding inherited properties like .constructor or .toString
[C,lib/utils/customs.js,52]       if (Object.hasOwnProperty.call(object, prop)) {
[C,lib/utils/customs.js,53]         return object[prop] = value;
[C,lib/utils/customs.js,54]       }
[C,lib/utils/customs.js,55]     }
[C,lib/utils/customs.js,56]     else {
[C,lib/utils/customs.js,57]       // this is a new property, that's just ok
[C,lib/utils/customs.js,58]       return object[prop] = value;
[C,lib/utils/customs.js,59]     }
[C,lib/utils/customs.js,60]   }
[C,lib/utils/customs.js,61] 
[C,lib/utils/customs.js,62]   throw new Error('No access to property "' + prop + '"');
[C,lib/utils/customs.js,63] }
[C,lib/utils/customs.js,64] 
[A,lib/utils/customs.js,65] /**
[A,lib/utils/customs.js,66]  * Test whether a property is safe to use for an object.
[A,lib/utils/customs.js,67]  * For example .toString and .constructor are not safe
[A,lib/utils/customs.js,68]  * @param {string} prop
[A,lib/utils/customs.js,69]  * @return {boolean} Returns true when safe
[A,lib/utils/customs.js,70]  */
[A,lib/utils/customs.js,71] function isSafeProperty (prop) {
[A,lib/utils/customs.js,72]   return !(prop in {});
[A,lib/utils/customs.js,73] }
[A,lib/utils/customs.js,74] 
[C,lib/utils/customs.js,75] /**
[C,lib/utils/customs.js,76]  * Validate whether a method is safe.
[C,lib/utils/customs.js,77]  * Throws an error when that's not the case.
[C,lib/utils/customs.js,78]  * @param {Object} object
[C,lib/utils/customs.js,79]  * @param {string} method
[C,lib/utils/customs.js,80]  */
[C,lib/utils/customs.js,81] // TODO: merge this function into assign.js?
[C,lib/utils/customs.js,82] function validateSafeMethod (object, method) {
[C,lib/utils/customs.js,83]   if (!isSafeMethod(object, method)) {
[C,lib/utils/customs.js,84]     throw new Error('No access to method "' + method + '"');
[C,lib/utils/customs.js,85]   }
[C,lib/utils/customs.js,86] }
[C,lib/utils/customs.js,87] 
[C,lib/utils/customs.js,88] function isSafeMethod (object, method) {
[C,lib/utils/customs.js,89]   // TODO: remove this, replace with whitelist
@@ -355,17 +365,18 @@ var safeMethods = {
[C,lib/utils/customs.js,365]   typeof: true,
[C,lib/utils/customs.js,366]   unaryMinus: true,
[C,lib/utils/customs.js,367]   unaryPlus: true,
[C,lib/utils/customs.js,368]   unequal: true,
[C,lib/utils/customs.js,369]   unit: true,
[C,lib/utils/customs.js,370]   usolve: true,
[C,lib/utils/customs.js,371]   valueOf: true,
[C,lib/utils/customs.js,372]   var: true,
[C,lib/utils/customs.js,373]   xgcd: true,
[C,lib/utils/customs.js,374]   xor: true,
[C,lib/utils/customs.js,375]   zeros: true
[C,lib/utils/customs.js,376] }
[C,lib/utils/customs.js,377] 
[C,lib/utils/customs.js,378] exports.getSafeProperty = getSafeProperty;
[C,lib/utils/customs.js,379] exports.setSafeProperty = setSafeProperty;
[A,lib/utils/customs.js,380] exports.isSafeProperty = isSafeProperty;
[C,lib/utils/customs.js,381] exports.validateSafeMethod = validateSafeMethod;
[C,lib/utils/customs.js,382] exports.isSafeMethod = isSafeMethod;
@@ -224,30 +224,39 @@ exports.traverse = function(object, path) {
[C,lib/utils/object.js,224] 
[C,lib/utils/object.js,225]   if (path) {
[C,lib/utils/object.js,226]     var names = path.split('.');
[C,lib/utils/object.js,227]     for (var i = 0; i < names.length; i++) {
[C,lib/utils/object.js,228]       var name = names[i];
[C,lib/utils/object.js,229]       if (!(name in obj)) {
[C,lib/utils/object.js,230]         obj[name] = {};
[C,lib/utils/object.js,231]       }
[C,lib/utils/object.js,232]       obj = obj[name];
[C,lib/utils/object.js,233]     }
[C,lib/utils/object.js,234]   }
[C,lib/utils/object.js,235] 
[C,lib/utils/object.js,236]   return obj;
[C,lib/utils/object.js,237] };
[C,lib/utils/object.js,238] 
[A,lib/utils/object.js,239] /**
[A,lib/utils/object.js,240]  * A safe hasOwnProperty
[A,lib/utils/object.js,241]  * @param {Object} object
[A,lib/utils/object.js,242]  * @param {string} property
[A,lib/utils/object.js,243]  */
[A,lib/utils/object.js,244] exports.hasOwnProperty = function (object, property) {
[A,lib/utils/object.js,245]   return object && Object.hasOwnProperty.call(object, property);
[A,lib/utils/object.js,246] }
[A,lib/utils/object.js,247] 
[C,lib/utils/object.js,248] /**
[C,lib/utils/object.js,249]  * Test whether an object is a factory. a factory has fields:
[C,lib/utils/object.js,250]  *
[C,lib/utils/object.js,251]  * - factory: function (type: Object, config: Object, load: function, typed: function [, math: Object])   (required)
[C,lib/utils/object.js,252]  * - name: string (optional)
[C,lib/utils/object.js,253]  * - path: string    A dot separated path (optional)
[C,lib/utils/object.js,254]  * - math: boolean   If true (false by default), the math namespace is passed
[C,lib/utils/object.js,255]  *                   as fifth argument of the factory function
[C,lib/utils/object.js,256]  *
[C,lib/utils/object.js,257]  * @param {*} object
[C,lib/utils/object.js,258]  * @returns {boolean}
[C,lib/utils/object.js,259]  */
[C,lib/utils/object.js,260] exports.isFactory = function (object) {
[C,lib/utils/object.js,261]   return object && typeof object.factory === 'function';
[C,lib/utils/object.js,262] };
@@ -98,30 +98,44 @@ describe('security', function () {
[C,test/expression/security.test.js,98] 
[C,test/expression/security.test.js,99]   it ('should not allow calling Function via a symbol', function () {
[C,test/expression/security.test.js,100]     assert.throws(function () {
[C,test/expression/security.test.js,101]       math.eval('O = {}.constructor\n' +
[C,test/expression/security.test.js,102]           'd = O.getOwnPropertyDescriptor(O.__proto__, "constructor")\n' +
[C,test/expression/security.test.js,103]           'eval("value", d)("console.log(\'hacked...\')")()');
[C,test/expression/security.test.js,104]     }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,105]   })
[C,test/expression/security.test.js,106] 
[C,test/expression/security.test.js,107]   it ('should not allow calling Function via a specially encoded constructor property name', function () {
[C,test/expression/security.test.js,108]     assert.throws(function () {
[C,test/expression/security.test.js,109]       math.eval('[].map["\\x63onstructor"]("console.log(\'hacked...\')")()');
[C,test/expression/security.test.js,110]     }, /Error: Cannot access method "map" as a property/);
[C,test/expression/security.test.js,111]   })
[C,test/expression/security.test.js,112] 
[A,test/expression/security.test.js,113]   it ('should not allow creating an Object with forbidden properties', function () {
[A,test/expression/security.test.js,114]     assert.throws(function () {
[A,test/expression/security.test.js,115]       math.eval('{hasOwnProperty: 2}');
[A,test/expression/security.test.js,116]     }, /Error: No access to property "hasOwnProperty/);
[A,test/expression/security.test.js,117] 
[A,test/expression/security.test.js,118]     assert.throws(function () {
[A,test/expression/security.test.js,119]       math.eval('{constructor: 2}');
[A,test/expression/security.test.js,120]     }, /Error: No access to property "constructor/);
[A,test/expression/security.test.js,121] 
[A,test/expression/security.test.js,122]     assert.throws(function () {
[A,test/expression/security.test.js,123]       math.eval('{toString: 2}');
[A,test/expression/security.test.js,124]     }, /Error: No access to property "toString/);
[A,test/expression/security.test.js,125]   })
[A,test/expression/security.test.js,126] 
[C,test/expression/security.test.js,127]   it ('should not allow calling Object via a an object constructor', function () {
[C,test/expression/security.test.js,128]     assert.throws(function () {
[C,test/expression/security.test.js,129]             math.eval('{}.constructor.assign(expression.node.AssignmentNode.prototype, ' +
[C,test/expression/security.test.js,130]                 '{_compile: "".toString.bind("console.log(\'hacked...\')")})\n' +
[C,test/expression/security.test.js,131]                 'eval("a = 2")');
[C,test/expression/security.test.js,132]     }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,133]   })
[C,test/expression/security.test.js,134] 
[C,test/expression/security.test.js,135]   it ('should not allow calling Object via a __defineGetter__', function () {
[C,test/expression/security.test.js,136]     assert.throws(function () {
[C,test/expression/security.test.js,137]       math.eval('expression.node.AssignmentNode.prototype.__defineGetter__("_compile", ' +
[C,test/expression/security.test.js,138]           '{}.valueOf.bind("".toString.bind("console.log(\'hacked...\')"))); eval("a = 2")')
[C,test/expression/security.test.js,139]     }, /Error: Undefined symbol expression/);
[C,test/expression/security.test.js,140]   })
[C,test/expression/security.test.js,141] 
@@ -179,68 +193,68 @@ describe('security', function () {
[C,test/expression/security.test.js,193]     }, /Error: No access to property "value/);
[C,test/expression/security.test.js,194]   })
[C,test/expression/security.test.js,195] 
[C,test/expression/security.test.js,196]   it ('should not allow using restricted properties via subset (3)', function () {
[C,test/expression/security.test.js,197]     assert.throws(function () {
[C,test/expression/security.test.js,198]       math.eval('subset(parse("1"), index("value"), "eval").compile().eval()("console.log(\'hacked...\')")')
[C,test/expression/security.test.js,199]     }, /Error: No access to property "value/);
[C,test/expression/security.test.js,200]   })
[C,test/expression/security.test.js,201] 
[C,test/expression/security.test.js,202]   it ('should not allow inserting fake nodes with bad code via node.map or node.transform', function () {
[C,test/expression/security.test.js,203]     assert.throws(function () {
[C,test/expression/security.test.js,204]       math.eval("badValue = {\"isNode\": true, \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked\')\")")
[C,test/expression/security.test.js,205]     }, /Error: Cannot compile node: unknown type "undefined"/);
[C,test/expression/security.test.js,206] 
[C,test/expression/security.test.js,207]     assert.throws(function () {
[D,test/expression/security.test.js,194]       math.eval("badValue = {\"isNode\": true, \"type\": \"ConstantNode\", \"valueType\": \"string\", \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked\')\")")
[A,test/expression/security.test.js,208]       math.eval("badValue = {\"isNode\": true, \"type\": \"ConstantNode\", \"valueType\": \"string\", \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked...\')\")")
[C,test/expression/security.test.js,209]     }); // The error message is vague but well...
[C,test/expression/security.test.js,210]   })
[C,test/expression/security.test.js,211] 
[D,test/expression/security.test.js,198]   it.skip ('should not allow injecting code via an adjusted ConstantNode value', function () {
[A,test/expression/security.test.js,212]   it ('should not allow replacing validateSafeMethod with a local variant', function () {
[C,test/expression/security.test.js,213]     assert.throws(function () {
[D,test/expression/security.test.js,200]       // TODO: write unit test
[D,test/expression/security.test.js,201]       }, /Error: No access to property "constructor/);
[A,test/expression/security.test.js,214]       math.eval("eval(\"f(validateSafeMethod)=cos.constructor(\\\"return eval\\\")()\")(eval(\"f(x,y)=0\"))(\"console.log('hacked...')\")")
[A,test/expression/security.test.js,215]     }, /Error: Cannot compile node/);
[C,test/expression/security.test.js,216]   })
[C,test/expression/security.test.js,217] 
[D,test/expression/security.test.js,204]   it.skip ('should not allow injecting code via an adjusted SymbolNode name', function () {
[A,test/expression/security.test.js,218]   it ('should not allow abusing toString', function () {
[C,test/expression/security.test.js,219]     assert.throws(function () {
[D,test/expression/security.test.js,206]       // TODO: write unit test
[D,test/expression/security.test.js,207]       }, /Error: No access to property "constructor/);
[A,test/expression/security.test.js,220]       math.eval("badToString = eval(\"f() = 1\"); badReplace = eval(\"f(a, b) = \\\"eval\\\"\"); badNumber = {toString:badToString, replace:badReplace}; badNode = {\"isNode\": true, \"type\": \"ConstantNode\", \"valueType\": \"number\", \"value\": badNumber}; x = eval(\"f(child, path, parent) = badNode\", {badNode:badNode}); parse(\"(1)\").map(x).compile().eval()(\"console.log('hacked...')\")")
[A,test/expression/security.test.js,221]     }, /Error: No access to property "toString"/);
[C,test/expression/security.test.js,222]   })
[C,test/expression/security.test.js,223] 
[D,test/expression/security.test.js,210]   it.skip ('should not allow injecting code via an adjusted FunctionNode name', function () {
[A,test/expression/security.test.js,224]   it ('should not allow creating a bad FunctionAssignmentNode', function () {
[C,test/expression/security.test.js,225]     assert.throws(function () {
[D,test/expression/security.test.js,212]       // TODO: write unit test
[D,test/expression/security.test.js,213]       }, /Error: No access to property "constructor/);
[A,test/expression/security.test.js,226]       math.eval("badNode={isNode:true,type:\"FunctionAssignmentNode\",expr:parse(\"1\"),types:{join:eval(\"f(a)=\\\"\\\"\")},params:{\"forEach\":eval(\"f(x)=1\"),\"join\":eval(\"f(x)=\\\"){return eval;}});return fn;})())}});return fn;})());}};//\\\"\")}};parse(\"f()=x\").map(eval(\"f(a,b,c)=badNode\",{\"badNode\":badNode})).compile().eval()()()(\"console.log('hacked...')\")")
[A,test/expression/security.test.js,227]     }, /TypeError: fn is not a function/);
[C,test/expression/security.test.js,228]   })
[C,test/expression/security.test.js,229] 
[D,test/expression/security.test.js,216]   it.skip ('should not allow injecting code via an adjusted function name', function () {
[A,test/expression/security.test.js,230]   it ('should not allow creating a bad OperatorNode (1)', function () {
[C,test/expression/security.test.js,231]     assert.throws(function () {
[D,test/expression/security.test.js,218]         // TODO: write unit test
[D,test/expression/security.test.js,219]       }, /Error: No access to property "constructor/);
[A,test/expression/security.test.js,232]       math.eval("badNode={isNode:true,type:\"FunctionAssignmentNode\",expr:parse(\"1\"),types:{join:eval(\"f(a)=\\\"\\\"\")},params:{\"forEach\":eval(\"f(x)=1\"),\"join\":eval(\"f(x)=\\\"){return eval;}});return fn;})())}});return fn;})());}};//\\\"\")}};parse(\"f()=x\").map(eval(\"f(a,b,c)=badNode\",{\"badNode\":badNode})).compile().eval()()()(\"console.log('hacked...')\")")
[A,test/expression/security.test.js,233]     }, /TypeError: fn is not a function/);
[C,test/expression/security.test.js,234]   })
[C,test/expression/security.test.js,235] 
[D,test/expression/security.test.js,222]   it.skip ('should not allow injecting code via a property name', function () {
[A,test/expression/security.test.js,236]   it ('should not allow creating a bad OperatorNode (2)', function () {
[C,test/expression/security.test.js,237]     assert.throws(function () {
[D,test/expression/security.test.js,224]         // TODO
[D,test/expression/security.test.js,225]       }, /Error: No access to property "constructor/);
[A,test/expression/security.test.js,238]       math.eval("parse(\"(0)\").map(eval(\"f(a,b,c)=d\",{d:{isNode:true,type:\"OperatorNode\",fn:\"__lookupGetter__\",args:{map:eval(\"f(a)=b\",{b:{join:eval(\"f(a)=\\\"1)||eval;}};//\\\"\")}})}}})).compile().eval()(\"console.log('hacked...')\")")
[A,test/expression/security.test.js,239]     }, /TypeError: fn is not a function/);
[C,test/expression/security.test.js,240]   })
[C,test/expression/security.test.js,241] 
[D,test/expression/security.test.js,228]   it.skip ('should not allow inserting a fake ObjectNode', function () {
[A,test/expression/security.test.js,242]   it ('should not allow creating a bad ConstantNode', function () {
[C,test/expression/security.test.js,243]     assert.throws(function () {
[D,test/expression/security.test.js,230]       // TODO: write unit test
[D,test/expression/security.test.js,231]       }, /Error: No access to property "constructor/);
[A,test/expression/security.test.js,244]       math.eval('f(x,y)="eval";g()=3;fakeConstantNode={"isNode": true, "type": "ConstantNode", "valueType": "number", "value": {"replace": f, "toString": g}};injectFakeConstantNode(child,path,parent)=path=="value"?fakeConstantNode:child;parse("a=3").map(injectFakeConstantNode).compile().eval()("console.log(\'hacked...\')")')
[A,test/expression/security.test.js,245]     }, /Error: No access to property "toString"/);
[C,test/expression/security.test.js,246]   })
[C,test/expression/security.test.js,247] 
[C,test/expression/security.test.js,248]   it ('should allow calling functions on math', function () {
[C,test/expression/security.test.js,249]     assert.equal(math.eval('sqrt(4)'), 2);
[C,test/expression/security.test.js,250]   })
[C,test/expression/security.test.js,251] 
[C,test/expression/security.test.js,252]   it ('should allow accessing properties on an object', function () {
[C,test/expression/security.test.js,253]     assert.deepEqual(math.eval('obj.a', {obj: {a:42}}), 42);
[C,test/expression/security.test.js,254]   })
[C,test/expression/security.test.js,255] 
[C,test/expression/security.test.js,256]   it ('should not allow accessing inherited properties on an object', function () {
[C,test/expression/security.test.js,257]     assert.throws(function () {
[C,test/expression/security.test.js,258]       math.eval('obj.constructor', {obj: {a:42}});
[C,test/expression/security.test.js,259]     }, /Error: No access to property "constructor"/)
[C,test/expression/security.test.js,260]   })